### ISSUE TITLE
Extracting components using Stylus

### ISSUE THREAD

**Issue by olimorris** [tailwindcss/discuss#11]
Firstly, congratulations to @adamwathan, @reinink, @davidhemphill and Steve Schoger for an absolutely incredible release. Incredible in that it's immeasurably usable in it's current Alpha state and also that it's just beautiful and rapid-fire to use. Serious kudos to you all

My go to preprocessor for CSS is Stylus. Hooking up Tailwind with Stylus was absolutely no issue whatsoever until I came to trying out `extracted components`.

Messing around with vertical aligning flex items, this did not seem to work:

```
.y-center {
  @apply .flex .flex-col .justify-center
}
```
but was resolved by using Stylus' `@css` literal:
```
@css {
    .y-center {
        @apply .flex .flex-col .justify-center
    }
}
```

I haven't tried this out with any other preprocessor but more of a head's up for anyone whose using Stylus as well.

With limited knowledge I think it's to do with how Stylus processes [@rules](http://stylus-lang.com/docs/atrules.html). I note in my output `app.css` file, the `@apply` rule appears literally, whereas when I use the `@css` literal, it appears as expected.

**Comment by mehranrasulian** [comment]
@olimorris Thanks for sharing this.
But how about the situation when we want to use a `state variant` like **hover** with the `@apply` directive?

So if I have this style:
```javascript
@css {
  .btn {
    @apply .hover\:shaddow
  }
  // OR like this
  .btn {
    @apply hover:shaddow
  }
}
```

It gives me this error

```bash
`@apply` cannot be used with `.hover\\\:shaddow` because `.hover\\\:shaddow` either cannot be found, or it's actual definition includes a pseudo-selector like :hover, :active, etc.
```

**Comment by adamwathan** [comment]
Apply the regular utility in the pseudo selector, we talk about this in the
docs:

https://tailwindcss.com/docs/functions-and-directives#apply

**Comment by MartinMuzatko** [comment]
Same issue here. I also have trouble using `apply`.
using `@css` is not really an alternative here :/
What is the proposed solution? First run postcss and then stylus?

**Comment by adamwathan** [comment]
> using @css is not really a great alternative :/

Why not?

> What is the proposed solution? First run postcss and then stylus?

No, always run style first.

> Does stylus have a built-in function to include styles of other classes?

No, only Less supports this out of the box.

**Comment by MartinMuzatko** [comment]
With `@css` I need to resign from using the shortcuts the preprocessor offers to me.
Why use a pre-processor then at all, if all my `.styl` files consist of `@css {}` blocks.

So no.. `@css` is at its best a dirty workaround, where I practically disable the pre-processor.
If I want to mix `@apply` together with nested rules (based on stylus syntax, not `post-css` `@nest`), or use mixins, variables etc. I need to at least double the css class definitions. So now I have my components css scattered, instead of one place.

**Comment by adamwathan** [comment]
Unfortunately that's a Stylus problem, nothing we can do about that. Personally I just use `postcss-preset-env` to get features like nesting, and don't use any preprocessors at all.

**Comment by MartinMuzatko** [comment]
Of course it is not a problem of tailwind, I just hoped there were recommendations, other than that what was already said in the mentioned issue.
Thanks @adamwathan

**Comment by acidjazz** [comment]
I'm messing with different solutions to help remedy this situation, so far i've come up with a plugin that's easier on the eyes [here](https://github.com/stylus/stylus/issues/2335#issuecomment-451084800)

```stylus
.kevin
  apply(bg-orange text-blue)
  border 20px solid blue
```

**Comment by acidjazz** [comment]
Hey everyone just a heads up i got a fork of stylus that properly renders `@apply` rules properly, give it a try please! @mehrunrasoli @olimorris 

https://github.com/acidjazz/stylus

**Comment by hybridvision** [comment]
Hey @acidjazz, thanks so much for your Stylus fork! I've been using it in a project for a few months now and it has been solid. It's so nice to have Tailwind + Stylus together! 👍

**Comment by acidjazz** [comment]
i'm glad its working for ya @hybridvision !
lets get this PR [merged](https://github.com/stylus/stylus/pull/2442)!

**Comment by tobias-kuendig** [comment]
So, I've been using tailwind today in a Vue project together with Stylus and added the following rule:

```styl
.modal
    @apply inset-0 fixed bg-white
```

Without noticing it at first, it looks like the apply rule is working as expected thanks to @acidjazz's PR!

Maybe it's time to update the documentation?

![image](https://user-images.githubusercontent.com/8600029/68752413-91dbef80-0603-11ea-98f7-db11da773f80.png)

**Comment by hybridvision** [comment]
Yes, good point - since @acidjazz's PR was merged into the Stylus core, this now works out of the box so this note in the documentation could be removed  🎉  The only requirement is using a Stylus version that is v0.54.6 or later...

The `@screen` directive still doesn't work in Stylus but there are workarounds and it's less common to need this.
